qualtran/bloqs/chemistry/trotter.py

```diff
@@ -0,0 +1,553 @@
+#  Copyright 2023 Google LLC
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+from functools import cached_property
+from typing import Dict, Optional, Set, Tuple
+
+import numpy as np
+from attrs import frozen
+from cirq_ft import TComplexity
+from cirq_ft.algos.qrom import QROM
+from cirq_ft.infra.bit_tools import float_as_fixed_width_int
+from numpy.typing import NDArray
+
+from qualtran import Bloq, BloqBuilder, Register, Signature, SoquetT
+from qualtran.bloqs.arithmetic import (
+    Add,
+    MultiplyTwoReals,
+    OutOfPlaceAdder,
+    ScaleIntByReal,
+    SquareRealNumber,
+    SumOfSquares,
+)
+from qualtran.bloqs.basic_gates import Rz, TGate
+from qualtran.bloqs.basic_gates.rotation import RotationBloq
+from qualtran.cirq_interop import CirqGateAsBloq
+
+
+def get_polynomial_coeffs() -> Tuple[NDArray, NDArray]:
+    """Polynomial coefficients from Fusion paper.
+
+    Useful for tests and notebook so add as function.
+    """
+    poly_coeffs_a = np.array(
+        [
+            0.99994132489119882162,
+            0.49609891915903542303,
+            0.33261112772430493331,
+            0.14876762006038398086,
+        ]
+    )
+    poly_coeffs_b = np.array(
+        [
+            0.81648515205385221995,
+            0.27136515484240234115,
+            0.12756148214815175348,
+            0.044753028579153842218,
+        ]
+    )
+    return poly_coeffs_a, poly_coeffs_b
+
+
+def build_qrom_data_for_poly_fit(selection_bitsize: int, target_bitsize: int) -> NDArray:
+    """Build QROM data from polynomial coefficients from Fusion paper.
+
+    Args:
+        selection_bitsize: Number of bits for QROM selection register. This is
+            determined in practice by the number of bits required to store
+            r_{ij}^2.
+        target_bitsize: Number of bits of precision for polynomial coefficients.
+
+    Returns:
+        qrom_data: An array of integers representing the appropriately
+            repeated scaled fixed-point representation of the polynomial
+            coefficients required for the variable spaced QROM.
+    """
+    # We compute the inverse square root of x^2 using variable spaced QROM, interpolation and newton-raphson
+    # Build data so QROM recognizes repeated entries so as to use the variable spaced QROM implementation.
+    # The repeated ranges occur for l :-> l + 2^k, and are repeated twice
```

Comment:
`l :-> l + 2^k` what's this notation?

Reply:
Oh some poor notation copied from the reference. The idea is that the starting point for repeated data is spaced out by powers of 2, so [4, 5], [6, 7] (the data comes in pairs of two because we have two data sets), then [8, 11], [12, 15], ... this pattern repeats so that the data in the intervals [l, l+2^k] is repeated.